### ISSUE TITLE
 Get coverage data when using cross-compilation

### ISSUE THREAD

**Issue by Zzzzzz0H0** [bazelbuild/bazel#19205]
We're using bazel （version：5.1.1） as a build tool for our C++ software and currently we're running into some problems.
1. We have some cc_tests that need to get coverage on the arm platform, our toolchain does not support arm platform compilation now, but we can support cross compilation on x86.
2. We have remote execution and remote cache to speed up our build, and the remote worker supports both x86 and arm platforms. The x86 worker supports the build and running of the test case. The arm platform worker can only running but not build (because we do not have  arm platform toolchain)
3. When we need to run arm cc_test
    a. Cross compile on the x86 machine, and use rbe_x86, cache the build result to the remote cache
    b. Run the test case on the arm machine. Due to the caching mechanism of bazel, it will not build again at this time, runing the test case directly on the arm machine
4. now we need coverage of this cc_test, we cannot get the coverage use the same way.After executing bazel build, We use bazel coverage and bazel test --collect_code_coverage "--instrumentation_filter to try to running testcase and get coverage.  unfortunately, bazel still compiles some test programs and programs under bazel test/coverage. I cannot use bazel build Cache all the results needed to get the coverage

Bazel is a powerful build tool, but it seems to lack some flexibility in the face of this scenario, do you have some suggestions to solve the problem we are facing, thanks.

**Comment by c-mita** [comment]
I'm not sure I understand.

Code coverage typically requires targets to be rebuilt so the binaries can be instrumented. Are the relevant coverage flags are being passed during your "pre-cache" build step?

That is, are you passing `--collect_code_coverage` and `--instrumentation_filter` to the initial `blaze build` command?

**Comment by zhuheng-mark** [comment]
> That is, are you passing --collect_code_coverage and --instrumentation_filter to the initial blaze build command?

Yes, I added these 2 options to bazel build to try "pre-cache", the following is my bazel command and execution log.

I hope I can use a simple demo to explain the problem I am facing. I have a C++ lib called hello-time, and there is a test program called test-hello-time
`cc_library(
    name = "hello-time",
    srcs = ["hello-time.cc"],
    hdrs = ["hello-time.h"],
    # copts = ["-fno-PIC"],
    visibility = ["//main:__pkg__", "//test:__pkg__"],
)`

`cc_test(
    name = "test-hello-time",
    srcs = ["test-hello-time.cc"],
    deps = [
        "//lib:hello-time",
        "@gtest",
        "@gtest//:gtest_main",
    ],
)`


First, using remote  execution and cross-compilation try to "pre-cache":
`bazel build --subcommands --config=source --config=rbe  --config=aarch64    --collect_code_coverage   "--instrumentation_filter=^//..."   //test:test-hello-time &> aarch64_first_build.log`
note：**cofig=rbe**  indicate use  our x86  remote execution worker  and  remote storage; **--config=aarch64** indicate aarch65 target platform; **--subcommands** is convenient for me to see which commands have been executed

From the build log, we can see gcc  "Compiling lib/hello-time.cc" with "-fPIC -fprofile-arcs -ftest-coverage", "Compiling test/test-hello-time.cc" with "-fprofile-arcs -ftest-coverage"

Second, use bazel test to get coverage data:
`bazel test --subcommands --config=source --config=aarch64_rbe  --config=aarch64    --collect_code_coverage   "--instrumentation_filter=^//..."   //test:test-hello-time &> aarch64_secord_test.log `
note:  **--config=aarch64_rbe**   indicate use  our aarch64  remote execution worker  and **same remote storage with x86**

In this step, what I expect is to no longer execute gcc, but run test-hello-time directly. But from the test log, we can see "Compiling lib/hello-time.cc" with " -fprofile-arcs -ftest-coverage", and "Compiling test/test-hello-time.cc" with -fPIC -fprofile-arcs -ftest-coverage.  So in the end I get an exec format error. 

Then I did another test, I only executed bazle test on x86 platform ( also remove rbe config), from the log we can see 'Compiling lib/hello-time.cc' and 'Compiling test/test-hello -time.cc' appears 2 times, the difference between the two times is the option fPIC
`bazel test --subcommands --config=source   --collect_code_coverage   "--instrumentation_filter=^//..."   //test:test-hello-time &> x86_test_without_build.log`  
[aarch64_secord_test.log](https://github.com/bazelbuild/bazel/files/12344127/aarch64_secord_test.log)
[aarch64_first_build.log](https://github.com/bazelbuild/bazel/files/12344128/aarch64_first_build.log)
[x86_test_without_build.log](https://github.com/bazelbuild/bazel/files/12344129/x86_test_without_build.log)

I uploaded my log file, hope it can help you understand my testing process.
Hope to get further reply.

**Comment by c-mita** [comment]
Some of the command lines are different.
For instance, the compilation command to produce `test-hello-time.o` uses `-fPIC` in the "second_test.log" file (actually producing `test-hello-time.pic.o`) but not in the "first_build.log".

**Comment by zhuheng-mark** [comment]
Actually, I don't care about the specific parameter **-fPIC**. What I expect is that **bazel build** can cache all the compilation caches required by bazel test --collect_code_coverage **"--instrumentation_filter=^//..."** ,so that when I run  **bazel test** command after **bazel build,** there will be no any gcc command need to run (it can be done when remove the **--collect_code_coverage "--instrumentation_filter=^//..."** parameter)

**Comment by c-mita** [comment]
Yes, normally that would happen. But your various steps all use different configs; it's possible that there's some small difference.

The configuration matches on the compile actions, despite the difference in "-fPIC". But because of the error, I can't see what the linking step is. Can you run `aquery` on the test in question? It might offer a clue.

**Comment by zhuheng-mark** [comment]
fine, I run **aquery** commands:
`bazel aquery   --config=source   --collect_code_coverage   "--instrumentation_filter=^//..."   //test:test-hello-time`
[bazel-aquery.log](https://github.com/bazelbuild/bazel/files/12429888/bazel-aquery.log)
In the log we can see the input that generated coverage.dat
```
action 'Testing //test:test-hello-time'
  Mnemonic: TestRunner
  Target: //test:test-hello-time
  Configuration: k8-opt
  Execution platform: @local_config_platform//:host
  ActionKey: 7d036e7c092cbf97c0efa710aad673270886bd56770d4ecb16d3d754deab0642
  Inputs: [bazel-out/k8-opt/bin/external/remote_coverage_tools/Main, bazel-out/k8-opt/bin/external/remote_coverage_tools/Main.jar, bazel-out/k8-opt/bin/lib/_objs/hello-time/hello-time.gcno, bazel-out/k8-opt/bin/lib/_objs/hello-time/hello-time.pic.gcno, bazel-out/k8-opt/bin/test/_objs/test-hello-time/test-hello-time.gcno, bazel-out/k8-opt/bin/test/_objs/test-hello-time/test-hello-time.pic.gcno, bazel-out/k8-opt/bin/test/test-hello-time.instrumented_files, bazel-out/k8-opt/internal/_middlemen/external_Sremote_Ucoverage_Utools_SMain-runfiles, bazel-out/k8-opt/internal/_middlemen/test_Stest-hello-time-runfiles, external/bazel_tools/tools/test/collect_cc_coverage.sh, external/bazel_tools/tools/test/collect_coverage.sh, external/bazel_tools/tools/test/generate-xml.sh, external/bazel_tools/tools/test/test-setup.sh, external/local_config_cc/builtin_include_directory_paths]
  Outputs: [bazel-out/k8-opt/testlogs/test/test-hello-time/coverage.dat, bazel-out/k8-opt/testlogs/test/test-hello-time/test.cache_status, bazel-out/k8-opt/testlogs/test/test-hello-time/test.log]
  Environment: [INCEPTIO_IS_SOURCE=True]
  Command Line: (exec external/bazel_tools/tools/test/test-setup.sh \
    external/bazel_tools/tools/test/collect_coverage.sh \
    test/test-hello-time)
# Configuration: 9ba0de7862245c4dc8d08da0cdb912022fbc36e8b86fd0067382d81515d6a7d4
# Execution platform: @local_config_platform//:host
```

> bazel-out/k8-opt/bin/lib/_objs/hello-time/hello-time.gcno, bazel-out/k8-opt/bin/lib/_objs/hello-time/hello-time.pic.gcno, bazel-out/k8-opt/bin/test/_objs/test-hello-time/test-hello-time.gcno, bazel-out/k8-opt/bin/test/_objs/test-hello-time/test-hello-time.pic.gcno
**bazel test** needs both **hello-time** and **test-hello-time** **fPIC** and without **fPIC** versions to generate coverage files, but we can't cache all these files in **bazel build**

**Comment by zhuheng-mark** [comment]
> Alternatively, try testing with --features=-supports_pic to explicitly disable it?

I tried executing `bazel test --features=-supports_pic --subcommands --collect_code_coverage "--instrumentation_filter
=^//..." --config=source //test:test-hello-time` ,unfortunately the execution failed:
`/usr/bin/ld.gold: error: bazel-out/k8-opt/bin/lib/_objs/hello-time/hello-time.o: requires dynamic R_X86_64_PC32 reloc against '_ZSt4cout' which may overflow at runtime; recompile with -fPIC`

It seems that the **fPIC** option is necessary for  **cc_library**.

**Comment by c-mita** [comment]
The problem is that both PIC and non-PIC files are being generated by the compilation actions, but only one is being used by the link. So your build only generates one set (non-PIC in your case).

Unfortunately, coverage generation for the test action wants both sets. This can't really be avoided; a cc_library can't always guess ahead of time which of its compile outputs will be used, so it needs to provide both to `InstrumentedFilesInfo`.

So you need to ensure only one set is ever generated. If you can't disable PIC, then I think the only way you can do this  is via forcing it using `--force_pic`. This should prevent your compile actions from generating non-PIC objects and ensure the link action will use the PIC objects.
